{"cells":[{"cell_type":"code","execution_count":null,"metadata":{"id":"OpFohZMCNtXh"},"outputs":[],"source":["%matplotlib inline"]},{"cell_type":"markdown","metadata":{"id":"vX6pXkNENtXq"},"source":["\n","# DL2Vec\n","\n","This example corresponds to the paper [Predicting candidate genes from phenotypes, functions and anatomical site of expression](https://doi.org/10.1093/bioinformatics/btaa879).\n","\n","This work is a graph-based machine-learning method to learn from biomedical ontologies. This method works by transforming the ontology into a graph following a set of rules. Random walks are generated from the obtained graph and then processed by a Word2Vec model, which generates\n","embeddings of the original ontology classes. This algorithm is applied to generate numerical representations of genes and diseases based on the background knowledge found in the Gene Ontology, which was extended to incorporate phenotypes, functions of the gene products and anatomical location of gene expression. The representations of genes and diseases are then used to predict candidate genes for a given disease.\n"]},{"cell_type":"markdown","metadata":{"id":"9DyOjZ4uNtXu"},"source":["To show an example of DL2Vec, we need 3 components:\n","\n","- The ontology projector\n","- The random walks generator\n","- The Word2Vec model\n","\n"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"1wcsmHQrNtXw"},"outputs":[],"source":["import sys\n","sys.path.append('../../')\n","import mowl\n","mowl.init_jvm(\"10g\")\n","\n","from mowl.datasets.builtin import GDAMouseDataset\n","from mowl.projection import DL2VecProjector\n","from mowl.walking import DeepWalk\n","from gensim.models.word2vec import LineSentence\n","from gensim.models import Word2Vec"]},{"cell_type":"markdown","metadata":{"id":"awFEksNwNtXx"},"source":["## Projecting the ontology\n","\n","We project the ontology using the DL2VecProjector class. The rules used to project the\n","ontology can be found at :doc:`/graphs/projection`. The outcome of the projection algorithm\n","is an edgelist.\n","\n"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"zPFKhxIgNtXz"},"outputs":[],"source":["dataset = GDAMouseDataset()\n","\n","projector = DL2VecProjector(bidirectional_taxonomy=True)\n","edges = projector.project(dataset.ontology)"]},{"cell_type":"markdown","metadata":{"id":"I_hy7_kmNtX0"},"source":["## Generating random walks\n","\n","The random walks are generated using the DeepWalk class. This class implements the DeepWalk\n","algorithm with a modification consisting of including the edge labels as part of the walks.\n","\n"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"vwGmTMlcNtX1"},"outputs":[],"source":["walker = DeepWalk(20, # number of walks per node\n","                  20, # walk length\n","                  0.1, # restart probability\n","                  workers=4) # number of threads\n","\n","walks = walker.walk(edges)"]},{"cell_type":"markdown","metadata":{"id":"KCdqXcVYNtX2"},"source":["## Training the Word2Vec model\n","\n","To train the Word2Vec model, we rely on the Gensim library:\n","\n"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"lBkdInxONtX3"},"outputs":[],"source":["walks_file = walker.outfile\n","sentences = LineSentence(walks_file)\n","model = Word2Vec(sentences, vector_size=100, epochs = 20, window=5, min_count=1, workers=4)"]},{"cell_type":"markdown","metadata":{"id":"NduBDv40NtX4"},"source":["## Evaluating the embeddings\n","\n","We can evaluate the embeddings using the\n",":class:`EmbeddingsRankBasedEvaluator <mowl.evaluation.rank_based.EmbeddingsRankBasedEvaluator>`\n","class. We need to do some data preparation.\n","\n"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"uyblHHdmNtX5"},"outputs":[],"source":["from mowl.evaluation.rank_based import EmbeddingsRankBasedEvaluator\n","from mowl.evaluation.base import CosineSimilarity\n","from mowl.projection import TaxonomyWithRelationsProjector"]},{"cell_type":"markdown","metadata":{"id":"LWFZWAtQNtX6"},"source":["We are going to evaluate the plausability of an association gene-disease with a gene against all\n","possible diseases and check the rank of the true disease association.\n","\n"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"bJJgrdtGNtX6"},"outputs":[],"source":["genes, diseases = dataset.evaluation_classes\n","\n","projector = TaxonomyWithRelationsProjector(taxonomy=False,\n","                                           relations=[\"http://is_associated_with\"])\n","\n","evaluation_edges = projector.project(dataset.testing)\n","filtering_edges = projector.project(dataset.ontology)\n","assert len(evaluation_edges) > 0"]},{"cell_type":"markdown","metadata":{"id":"s-riM6ClNtX7"},"source":["The gene-disease associations will be scoredc using cosine similarity. For that reason we use\n","the ``CosineSimilarity`` class.\n","\n"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"3heluKb1NtX7"},"outputs":[],"source":["vectors = model.wv\n","evaluator = EmbeddingsRankBasedEvaluator(\n","    vectors,\n","    evaluation_edges,\n","    CosineSimilarity,\n","    training_set=filtering_edges,\n","    head_entities = genes.as_str,\n","    tail_entities = diseases.as_str,\n","    device = 'cpu'\n",")\n","\n","evaluator.evaluate(show=True)"]}],"metadata":{"kernelspec":{"display_name":"Python 3","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.8.16"},"colab":{"provenance":[]}},"nbformat":4,"nbformat_minor":0}